src/Bicep.Core/Semantics/NameBindingVisitor.cs

```diff
@@ -1,341 +1,410 @@
+            this.activeScopes = new Stack<LocalScope>();
+        }
+
+        public override void VisitProgramSyntax(ProgramSyntax syntax)
+        {
+            base.VisitProgramSyntax(syntax);
+
+            // create bindings for all of the declarations to their corresponding symbol
+            // this is needed to make find all references work correctly
+            // (doing this here to avoid side-effects in the constructor)
+            foreach (DeclaredSymbol declaredSymbol in this.declarations.Values)
+            {
+                this.bindings.Add(declaredSymbol.DeclaringSyntax, declaredSymbol);
+            }
+
+            // include all the locals in the symbol table as well
+            // since we only allow lookups by object and not by name,
+            // a flat symbol table should be sufficient
+            foreach (var declaredSymbol in allLocalScopes.Values.SelectMany(scope => scope.Declarations))
+            {
+                this.bindings.Add(declaredSymbol.DeclaringSyntax, declaredSymbol);
+            }
+        }
+
+        public override void VisitVariableAccessSyntax(VariableAccessSyntax syntax)
+        {
+            base.VisitVariableAccessSyntax(syntax);
+
+            var symbol = this.LookupSymbolByName(syntax.Name, false);
+
+            // bind what we got - the type checker will validate if it fits
+            this.bindings.Add(syntax, symbol);
+        }
+
+        public override void VisitResourceAccessSyntax(ResourceAccessSyntax syntax)
+        {
+            base.VisitResourceAccessSyntax(syntax);
+
+            // we need to resolve which resource delaration the LHS is pointing to - and then 
+            // validate that we can resolve the name.
+            this.bindings.TryGetValue(syntax.BaseExpression, out var symbol);
+
+            if (symbol is ErrorSymbol)
+            {
+                this.bindings.Add(syntax, symbol);
+                return;
+            }
+            else if (symbol is null || symbol is not ResourceSymbol)
+            {
+                // symbol could be null in the case of an incomplete expression during parsing like `a:`
+                var error = new ErrorSymbol(DiagnosticBuilder.ForPosition(syntax.ResourceName).ResourceRequiredForResourceAccess(symbol?.Kind.ToString() ?? LanguageConstants.ErrorName));
+                this.bindings.Add(syntax, error);
+                return;
+            }
+
+            // This is the symbol of LHS and it's a valid resource.
+            var resourceSymbol = (ResourceSymbol)symbol;
+            var resourceBody = resourceSymbol.DeclaringResource.TryGetBody();
+            if (resourceBody == null)
+            {
+                // If we have no body then there will be nothing to reference.
+                var error = new ErrorSymbol(DiagnosticBuilder.ForPosition(syntax.ResourceName).ResourceRequiredForResourceAccess(symbol?.Kind.ToString() ?? LanguageConstants.ErrorName));

```

Comment:
If everything is valid but you try an incorrect resource access, you get the following message (which is great):
![image](https://user-images.githubusercontent.com/22460039/109704631-ab33e280-7b4b-11eb-8a03-f01f3423c222.png)

But if the resource body is not valid you get the following:
![image](https://user-images.githubusercontent.com/22460039/109704869-f5b55f00-7b4b-11eb-9f1c-3d5bc9b69fda.png)

I know what you were going for, but I think it'll be misleading. (I'm referencing a resource. Why isn't it a resource type?) I think a better working would be something like this (or something similar):
`The resource "<whatever>" does not contain a nested resource named "<whatever2>. There are no available nested resources under resource "<whatever>".`

Reply:
@majastrz - can you attach the case you used to reproduce that error? 

The error message you're referring to is **intended** for the case where we cannot determine the resource type of the parent, and thus cannot determine what children are valid. it sounds like the error in the resource body has an unintended effect.

Reply:
Agreed that based on your description this should be better.

Reply:
It was this:
```bicep
resource broken 'Microsoft.Network/virtualNetworks@2020-06-01' = 

output foo string = broken:test
```

Reply:
Fixed this, it was indeed the wrong error message for this case.